### PR TITLE
Fixed error "ValueError: invalid literal for int() with base 10: 'undefined'"

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
fixes https://github.com/mitodl/edx-sga/issues/159#issue-205215793

### For testing:
-  Follow steps posted on issue.
- Change this [line in edX platform](https://github.com/edx/edx-platform/blob/master/requirements/edx/github.txt#L96) to ```-e git+https://github.com/mitodl/edx-sga.git@fix/aq/exceptions_base_10#egg=edx-sga```
 
@pdpinch @singingwolfboy

### CODE REVIEW

**Comment by pdpinch** [comment]
I think I've got your code installed correctly -- at least, I'm not getting the `ValueError` error,  but now I'm getting a different set of errors. 

I think some of these errors are due to several ajax requests being fired (possibly fixed by PR #131 ?), but I don't know what to make of the `TransactionManagementError`

I'm running this with current master, using a ficus devstack. 

```
2017-02-05 19:12:38,895 ERROR 30624 [submissions.api] api.py:765 - Error occurred while reseting scores for item block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@1c6b1263fa8546cfa42b4a2708f2c929 in course course-v1:edX+DemoX+Demo_Course for student 85659512ec6b816edc95564b8ceb9d03
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/src/edx-submissions/submissions/api.py", line 747, in reset_score
    created_at=score.created_at,
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/signals/handlers.py", line 103, in submissions_score_reset_handler
    score_db_table=ScoreDatabaseTableEnum.submissions,
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/signals/handlers.py", line 195, in enqueue_subsection_update
    countdown=RECALCULATE_GRADE_DELAY,
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py", line 551, in apply_async
    link=link, link_error=link_error, **options)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py", line 743, in apply
    request=request, propagate=throw)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 355, in eager_trace_task
    uuid, args, kwargs, request)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 253, in trace_task
    I, R, state, retval = on_error(task_request, exc, uuid)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 201, in on_error
    R = I.handle_error_state(task, eager=eager)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 85, in handle_error_state
    }[self.state](task, store_errors=store_errors)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 120, in handle_failure
    task.on_failure(exc, req.id, req.args, req.kwargs, einfo)
  File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/celery_utils/persist_on_failure.py", line 19, in on_failure
    if not FailedTask.objects.filter(task_id=task_id, datetime_resolved=None).exists():
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/query.py", line 586, in exists
    return self.query.has_results(using=self.db)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/sql/query.py", line 484, in has_results
    return compiler.has_results()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 811, in has_results
    return bool(self.execute_sql(SINGLE))
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 840, in execute_sql
    cursor.execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/debug_toolbar/panels/sql/tracking.py", line 159, in execute
    return self._record(self.cursor.execute, sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/debug_toolbar/panels/sql/tracking.py", line 101, in _record
    return method(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 59, in execute
    self.db.validate_no_broken_transaction()
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/db/backends/base/base.py", line 327, in validate_no_broken_transaction
    "An error occurred in the current transaction. You can't "
TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.
2017-02-05 19:12:38,896 ERROR 30624 [courseware.module_render] module_render.py:1009 - error executing xblock handler
Traceback (most recent call last):
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 984, in _invoke_xblock_handler
    resp = instance.handle(handler, req, suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 86, in handle
    return self.runtime.handle(self, handler_name, request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1339, in handle
    return super(MetricsMixin, self).handle(block, handler_name, request, suffix=suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 1025, in handle
    results = handler(request, suffix)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-sga/edx_sga/sga.py", line 445, in remove_grade
    submissions_api.reset_score(student_id, self.course_id, self.block_id)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-submissions/submissions/api.py", line 766, in reset_score
    raise SubmissionInternalError(msg)
SubmissionInternalError: Error occurred while reseting scores for item block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@1c6b1263fa8546cfa42b4a2708f2c929 in course course-v1:edX+DemoX+Demo_Course for student 85659512ec6b816edc95564b8ceb9d03
2017-02-05 19:12:38,899 ERROR 30624 [root] exceptions.py:16 - Uncaught exception from <class 'django.core.handlers.wsgi.WSGIHandler'>
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 877, in handle_xblock_callback
    return _invoke_xblock_handler(request, course_id, usage_id, handler, suffix, course=course)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 984, in _invoke_xblock_handler
    resp = instance.handle(handler, req, suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 86, in handle
    return self.runtime.handle(self, handler_name, request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1339, in handle
    return super(MetricsMixin, self).handle(block, handler_name, request, suffix=suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 1025, in handle
    results = handler(request, suffix)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-sga/edx_sga/sga.py", line 445, in remove_grade
    submissions_api.reset_score(student_id, self.course_id, self.block_id)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-submissions/submissions/api.py", line 766, in reset_score
    raise SubmissionInternalError(msg)
SubmissionInternalError: Error occurred while reseting scores for item block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@1c6b1263fa8546cfa42b4a2708f2c929 in course course-v1:edX+DemoX+Demo_Course for student 85659512ec6b816edc95564b8ceb9d03
Internal Server Error: /courses/course-v1:edX+DemoX+Demo_Course/xblock/block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@1c6b1263fa8546cfa42b4a2708f2c929/handler/remove_grade
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 877, in handle_xblock_callback
    return _invoke_xblock_handler(request, course_id, usage_id, handler, suffix, course=course)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 984, in _invoke_xblock_handler
    resp = instance.handle(handler, req, suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 86, in handle
    return self.runtime.handle(self, handler_name, request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1339, in handle
    return super(MetricsMixin, self).handle(block, handler_name, request, suffix=suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 1025, in handle
    results = handler(request, suffix)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-sga/edx_sga/sga.py", line 445, in remove_grade
    submissions_api.reset_score(student_id, self.course_id, self.block_id)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-submissions/submissions/api.py", line 766, in reset_score
    raise SubmissionInternalError(msg)
SubmissionInternalError: Error occurred while reseting scores for item block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@1c6b1263fa8546cfa42b4a2708f2c929 in course course-v1:edX+DemoX+Demo_Course for student 85659512ec6b816edc95564b8ceb9d03
2017-02-05 19:12:38,900 ERROR 30624 [django.request] base.py:256 - Internal Server Error: /courses/course-v1:edX+DemoX+Demo_Course/xblock/block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@1c6b1263fa8546cfa42b4a2708f2c929/handler/remove_grade
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 877, in handle_xblock_callback
    return _invoke_xblock_handler(request, course_id, usage_id, handler, suffix, course=course)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 984, in _invoke_xblock_handler
    resp = instance.handle(handler, req, suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 86, in handle
    return self.runtime.handle(self, handler_name, request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1339, in handle
    return super(MetricsMixin, self).handle(block, handler_name, request, suffix=suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 1025, in handle
    results = handler(request, suffix)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-sga/edx_sga/sga.py", line 445, in remove_grade
    submissions_api.reset_score(student_id, self.course_id, self.block_id)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-submissions/submissions/api.py", line 766, in reset_score
    raise SubmissionInternalError(msg)
SubmissionInternalError: Error occurred while reseting scores for item block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@1c6b1263fa8546cfa42b4a2708f2c929 in course course-v1:edX+DemoX+Demo_Course for student 85659512ec6b816edc95564b8ceb9d03
[05/Feb/2017 19:12:39] "GET /courses/course-v1:edX+DemoX+Demo_Course/xblock/block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@1c6b1263fa8546cfa42b4a2708f2c929/handler/remove_grade?module_id=11&student_id=85659512ec6b816edc95564b8ceb9d03 HTTP/1.1" 500 86054
```

**Comment by amir-qayyum-khan** [comment]
May be because i uses `Eucalyptus` for testing.

**Comment by pdpinch** [comment]
Ok. I'll try with Eucalyptus.

**Comment by pdpinch** [comment]
Can you export your course to a zip file & send it to me?

I reverted to Eucalyptus (still using master) and I'm still getting this error. I did notice that I'm testing with a staff account, trying to remove my own score. Maybe that's what leads to the `TransactionManagementError` error. 

In Eucalyptus, when I test with a student submission, I get the following error:

```
2017-02-06 21:39:13,661 ERROR 4728 [django.request] base.py:256 - Internal Server Error: /courses/course-v1:edX+DemoX+Demo_Course/xblock/block-v1:edX+DemoX+Demo_Course+type@edx_sga+block@3c18c6f7fde04a5da56d6853b4169e87/handler/remove_grade
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 877, in handle_xblock_callback
    return _invoke_xblock_handler(request, course_id, usage_id, handler, suffix, course=course)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/courseware/module_render.py", line 984, in _invoke_xblock_handler
    resp = instance.handle(handler, req, suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/mixins.py", line 86, in handle
    return self.runtime.handle(self, handler_name, request, suffix)
  File "/edx/app/edxapp/edx-platform/common/lib/xmodule/xmodule/x_module.py", line 1339, in handle
    return super(MetricsMixin, self).handle(block, handler_name, request, suffix=suffix)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/xblock/runtime.py", line 1025, in handle
    results = handler(request, suffix)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-sga/edx_sga/sga.py", line 592, in remove_grade
    submissions_api.reset_score(student_id, unicode(self.course_id), self.block_id)
  File "/edx/app/edxapp/venvs/edxapp/src/edx-submissions/submissions/api.py", line 747, in reset_score
    created_at=score.created_at,
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/signals/handlers.py", line 103, in submissions_score_reset_handler
    score_db_table=ScoreDatabaseTableEnum.submissions,
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/edx/app/edxapp/edx-platform/lms/djangoapps/gating/signals.py", line 26, in handle_score_changed
    block = modulestore().get_item(UsageKey.from_string(kwargs.get('usage_id')))
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/opaque_keys/__init__.py", line 193, in from_string
    namespace, rest = cls._separate_namespace(serialized)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/opaque_keys/__init__.py", line 213, in _separate_namespace
    namespace, __, rest = serialized.partition(cls.NAMESPACE_SEPARATOR)
AttributeError: 'BlockUsageLocator' object has no attribute 'partition'
```

**Comment by amir-qayyum-khan** [comment]
Sent you email

**Comment by amir-qayyum-khan** [comment]
@pdpinch 

- fixed `AttributeError: 'CourseLocator' object has no attribute 'partition'`
- add server side validation
- tested SGA on ficus

**Comment by pdpinch** [comment]
Functionality is working now!

File: edx_sga/sga.py [hunk 1]
```diff
@@ -561,9 +561,21 @@ def enter_grade(self, request, suffix=''):
         Persist a score for a student given by staff.
         """
         require(self.is_course_staff())
+        score = request.params.get('grade', None)
         module = StudentModule.objects.get(pk=request.params['module_id'])
+        if not score:
+            log.error(
+                "[enter_grade] Please enter valid grade for course:%s module:%s student:%s",
```
Comment (pdpinch) [inline edx_sga/sga.py]: This is an odd error message for a log. Wouldn't something like `enter_grade: invalid grade submitted for course:%s module:%s student:%s` make more sense?

The square brackets aren't used anywhere else either.

File: edx_sga/sga.py [hunk 2]
```diff
@@ -561,9 +561,21 @@ def enter_grade(self, request, suffix=''):
...
+                module.module_state_key,
+                module.student.username
+            )
+            return Response(json_body={
+                "error": "Please enter valid grade"
+            })
+
         state = json.loads(module.state)
-        score = int(request.params['grade'])
+        score = int(score)
```
Comment (pdpinch) [inline edx_sga/sga.py]: What happens if we can't cast score into an int? Isn't that what started our problem?

File: edx_sga/static/js/src/edx_sga.js [hunk 3]
```diff
@@ -196,25 +207,29 @@ function StaffGradedAssignmentXBlock(runtime, element) {
                 var max_score = row.parents('#grade-info').data('max_score');
                 var score = Number(form.find('#grade-input').val());
                 event.preventDefault();
-                if (isNaN(score)) {
-                    form.find('.error').html('<br/>Grade must be a number.');
+                if (!score) {
```
Comment (pdpinch) [inline edx_sga/static/js/src/edx_sga.js]: Why did you change this from isNaN() to just ! ?

Comment (amir-qayyum-khan) [inline edx_sga/static/js/src/edx_sga.js]: i found that `isNaN` was not working in case of empty fields on chrome

File: edx_sga/sga.py (hunk 2 above)
Comment (amir-qayyum-khan) [inline edx_sga/sga.py]: i added exception handling on it

**Comment by amir-qayyum-khan** [comment]
Fixed @pdpinch

File: edx_sga/sga.py [hunk 4]
```diff
@@ -554,16 +554,45 @@ def get_staff_grading_data(self, request, suffix=''):
         require(self.is_course_staff())
         return Response(json_body=self.staff_grading_data())
 
+    def validat_score_message(self, course_id, username):
```
Comment (pdpinch) [inline edx_sga/sga.py]: I didn't see this earlier, but it should be `validate_score_message` 

You dropped an `e` at the end of `validate`

File: edx_sga/sga.py [hunk 5]
```diff
@@ -554,16 +554,45 @@ def get_staff_grading_data(self, request, suffix=''):
...
         # pylint: disable=unused-argument
         """
         Persist a score for a student given by staff.
         """
         require(self.is_course_staff())
+        score = request.params.get('grade', None)
         module = StudentModule.objects.get(pk=request.params['module_id'])
+        if not score:
+            return Response(
+                json_body=self.validat_score_message(
```
Comment (pdpinch) [inline edx_sga/sga.py]: `validate_score_message`

File: edx_sga/sga.py [hunk 6]
```diff
@@ -554,16 +554,45 @@ def get_staff_grading_data(self, request, suffix=''):
...
+                )
+            )
+
         state = json.loads(module.state)
-        score = int(request.params['grade'])
+        try:
+            score = int(score)
+        except ValueError:
+            return Response(
+                json_body=self.validat_score_message(
```
Comment (pdpinch) [inline edx_sga/sga.py]: `validate_score_message`
